### ISSUE TITLE
Feature Importance with objective:binary is double?

### ISSUE THREAD

**Issue by masaru-tsuruta** [microsoft/LightGBM#2708]
I have tried to confirm a value of the feature importance whose type is 'gain'.
When I have set 'objective' as 'binary', its value is about twice as much as change of log-loss.
(*Note that from my understanding, feature importance is calculated by using approximating with gradient and hessian in this setting. The value of Feature Importance is excactly twice as much as change of approximated log-loss which we calculate in Excel.)
In this way, 'gain(diff of logloss) ' is not equal to 'importance' of X. Is this correct?

(*Note that if we set 'objective' as 'regression' and use mse as score, these values are same.)

My code and result are as follows.
```
np.random.seed(seed=1)
df = pd.DataFrame([])
df['target'] = randint(0,2,4)  
df['X'] = randint(0,30,4) 

feat = list(['X'])
X, y = df.loc[:,feat].values,df.target.values

trn_data = lgb.Dataset(X, label=y, feature_name=feat)

params = {"objective" : 'binary', ## 'binary' 
        "boosting": 'gbdt', "max_depth": -1,
        "learning_rate":1, # to create simple model
        "num_leaves" : 2, # to create simple model
        'min_data_in_leaf': 1, # fit to small sample data
        'min_child_weight': 0.0,
        'max_bin': 50,
    }

# 1split model
clfx = lgb.train(params, trn_data, 1, verbose_eval=False)
    
    
params['min_data_in_leaf'] = 100 # To avoid split

# no split model
clfx2 = lgb.train(params, trn_data, 1, verbose_eval=False)

# calculate gain (logloss)
score = [log_loss(y,clfx.predict(X))*len(y),
         log_loss(y,clfx2.predict(X))*len(y)]

# show logloss
print('no-split logloss :'+'{:.4f}'.format(score[1]))
print('first-split logloss :'+'{:.4f}'.format(score[0]))
print('gain(diff of logloss) :'+'{:.4f}'.format((score[1]-score[0])))

imp = pd.DataFrame(clfx.feature_importance(importance_type='gain'),index = feat,columns=['importance'])

print("------")
print(imp)
```

no-split logloss :2.7726
first-split logloss :2.0367
gain(diff of logloss) :0.7359

------
   importance
X     1.33333

**Comment by guolinke** [comment]
the gain isn't simply the delta of logloss.
refer to https://github.com/microsoft/LightGBM/blob/master/src/treelearner/feature_histogram.hpp#L498-L506

and https://xgboost.readthedocs.io/en/latest/tutorials/model.html

**Comment by masaru-tsuruta** [comment]
Thank you for your support.
If the code consistent with xgboost tutorial, I think that there are some difference.

I can transform Line 505 in your link as following.
[https://github.com/microsoft/LightGBM/blob/master/src/treelearner/feature_histogram.hpp#L505](url)
I assume that l1 = 0, l2 = 0 and loss function is log-loss to simplify the formula.

-(2.0 * sg_l1 * output + (sum_hessians + l2) * output * output)  **(Line 505)**
= -(2.0 * sum_gradients + sum_hessians * output* output)
   (I assume that l1 = 0 and l2 = 0, and therefore sg_l1 = sum_gradients)
= -(2.0 * sum_gradients * ( - sum_gradients / sum_hessians) 
   + sum_hessians * ( - sum_gradients / sum_hessians) * ( - sum_gradients / sum_hessians) )
= -( -2.0 * sum_gradients* sum_gradients/sum_hessians +  sum_gradients * sum_gradients / sum_hessians )
     (where hessian of logloss is always positive)
= sum_gradients * sum_gradients / sum_hessians

This is twice as much as the formula of XGBoost tutorial as your link.
[https://xgboost.readthedocs.io/en/latest/tutorials/model.html](url)
obj*=-1/2\sum_{j=1}^{T}\frac{G_j^2}{H_j+\lambda}+\gammaT
Gain = -1/2 [ \frac{G_L^2}{H_L+\lambda} + \frac{G_R^2}{H_R+\lambda} 
             -  \frac{(G_L+G_R)^2}{H_L+H_R+\lambda}] -\gamma

Should it be multiplied 1/2.0 to calculate the gain which is defined in XGBoost tutorial?
